Helper/Data.php

```diff
@@ -914,8 +914,9 @@ public function getOneClickPaymentMethods($customerId, $storeId, $grandTotal)
 
             // check if contractType is supporting the selected contractType for OneClick payments
             $allowedContractTypes = $agreementData['contractTypes'];
+
             // RecurringType::ONECLICK is kept in the if block to still display tokens that were created before changes
-            if (in_array(RecurringType::ONECLICK, $allowedContractTypes) || in_array(Recurring::CARD_ON_FILE, $allowedContractTypes)) {
+            if (is_null($subType) || in_array(RecurringType::ONECLICK, $allowedContractTypes) || in_array($subType, $allowedContractTypes)) {
```

Comment:
Thank you for the review @michaelpaul ! 

Yes, this is intentional as in the backend model we are not passing the $subType, because we want to show all the saved payments for the customer, saved with CardOnFile and Subscription contract type.
However, in the frontend model we want to show only the saved payments with CardOnFile, therefore I moved the `Recurring::CARD_ON_FILE` in the frontend model.

Reply:
This `if` condition could also return true if you pass `Recurring::SUBSCRIPTION` (or any non-null value) as the `$subType` argument, because if `in_array(RecurringType::ONECLICK, $allowedContractTypes) ||` is evaluated to `true` the PHP interpreter skips ahead and goes into the body of the `if` block.
If you want to return only `CardOnFile` payments in the frontend model, wouldn't it be better to do something like `if (($subType && in_array($subType, $allowedContractTypes)) || in_array(RecurringType::ONECLICK, $allowedContractTypes)) {`?

Reply:
Added the comment in the code!